src/opendr/perception/object_detection_2d/retinaface/algorithm/PY_OP/cascade_refine.py

```diff
@@ -0,0 +1,353 @@
+                    num_bg = max(48, n_fg * int(1.0 / config.TRAIN.RPN_FG_FRACTION - 1))
+
+                bg_inds = np.where(labels == 0)[0]
+                # origin_num_bg = len(bg_inds)
+                if num_bg == 0:
+                    labels[bg_inds] = -1
+                elif len(bg_inds) > num_bg:
+                    # sort ohem scores
+
+                    if self.mode == 0:
+                        disable_inds = np.random.choice(bg_inds, size=(len(bg_inds) - num_bg), replace=False)
+                        labels[disable_inds] = -1
+                    else:
+                        neg_ohem_scores = fg_score[bg_inds]
+                        order_neg_ohem_scores = neg_ohem_scores.ravel().argsort()[::-1]
+                        sampled_inds = bg_inds[order_neg_ohem_scores[:num_bg]]
+                        labels[bg_inds] = -1
+                        labels[sampled_inds] = 0
+
+                if n_fg > 0:
+                    order0_labels = labels.reshape((1, A, -1)).transpose((0, 2, 1)).reshape((-1,))
+                    bbox_fg_inds = np.where(order0_labels > 0)[0]
+                    _anchor_weight[bbox_fg_inds, :] = 1.0
+                anchor_weight[ibatch] = _anchor_weight
+                valid_count[ibatch][0] = n_fg
+                labels_out[ibatch] = labels
+                bbox_target_out[ibatch] = new_bbox_target
+        #     else:  # FAST MODE
+        #         fg_score_t0 = cls_score_t0[ibatch, 1, :] - cls_score_t0[ibatch, 0, :]
+        #         sort_idx_t0 = np.argsort(fg_score_t0.flatten())[::-1][0:SEL_TOPK]
+        #         _bbox_pred_t0 = bbox_pred_t0[ibatch][sort_idx_t0]
+        #         _bbox_target_t0 = bbox_target_t0[ibatch][sort_idx_t0]
+        #         anchors_t0 = self.apply_bbox_pred(_bbox_pred_t0)
+        #         gt_anchors = self.apply_bbox_pred(_bbox_target_t0)
+        #         gt_label = {'gt_boxes': gt_anchors}
+        #         new_label_dict = self.assign_anchor_fpn(gt_label, anchors_t0, False, prefix=self.prefix)
+        #         labels = new_label_dict['%s_label' % self.prefix]
+        #         new_bbox_target = new_label_dict['%s_bbox_target' % self.prefix]
+        #         _anchor_weight = np.zeros((num_anchors, 1), dtype=np.float32)
+        #         fg_score = cls_score[ibatch, 1, :] - cls_score[ibatch, 0, :]
+        #         fg_inds = np.where(labels > 0)[0]
+        #         _labels = np.empty(shape=labels.shape, dtype=np.float32)
+        #         _labels.fill(-1)
+        #         _labels[sort_idx_idx] = labels
+        #
+        #         anchor_weight[ibatch] = _anchor_weight
+        #         valid_count[ibatch][0] = len(fg_inds)
+        #         labels_out[ibatch] = _labels
+        #         bbox_target_out[ibatch] = new_bbox_target
+        #
+        # for ind, val in enumerate([labels_out, bbox_target_out, anchor_weight, valid_count]):
+        #     val = mx.nd.array(val)
+        #     self.assign(out_data[ind], req[ind], val)
```

Comment:
I don't know what's has been the policy with regards to commented functionality, whether it should be removed or if it can stay. I just point it out in case you know.

Reply:
Hi Daniel, I'm also unsure about whether it can stay but in the end I decided on deleting all commented lines of code.